### ISSUE TITLE
`pl.concat(..., how="horizontal")` fails if any item are series

### ISSUE THREAD

**Issue by mcrumiller** [pola-rs/polars#14951]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
>>> import polars as pl
>>> s = pl.Series("s", [1, 2, 3])
>>> df = pl.DataFrame({"df": [4, 5, 6]})
>>> pl.concat((s, s), how="horizontal")
ValueError: Series only supports 'vertical' concat strategy
>>> pl.concat((s, df), how="horizontal")
ValueError: Series only supports 'vertical' concat strategy
>>> pl.concat((df, s), how="horizontal")
ValueError: Series only supports 'vertical' concat strategy
```

This seems like it should work, as a Series can be trivially converted to a one-column frame.

### Issue description

a

### Expected behavior

a

### Installed versions

<details>

```
Polars:               0.20.14
Index type:           UInt32
Platform:             Windows-10-10.0.19045-SP0
Python:               3.11.7 (tags/v3.11.7:fa7a6f2, Dec  4 2023, 19:24:49) [MSC v.1937 64 bit (AMD64)]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          <not installed>
connectorx:           0.3.2
deltalake:            <not installed>
fastexcel:            <not installed>
fsspec:               <not installed>
gevent:               <not installed>
hvplot:               <not installed>
matplotlib:           3.8.2
numpy:                1.26.2
openpyxl:             3.1.2
pandas:               2.1.4
pyarrow:              14.0.1
pydantic:             <not installed>
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           2.0.23
xlsx2csv:             0.8.2
xlsxwriter:           3.1.9
```

</details>

**Comment by orlp** [comment]
I'm marking this as an enhancement rather than as a bug since the docs clearly state that `pl.concat` only allows series in vertical mode.

**Comment by orlp** [comment]
> Is a named series a series whose name is not `""`?

Yes.

> How do we know that the series was not explicitly named this?

I don't really care to discuss the scenario where someone intentionally names their columns the empty string. The vast majority of unnamed series are due to people simply not naming them, which is not something I'd like to support in `concat`.

**Comment by mickvangelderen** [comment]
There is [`Series.to_frame`](https://docs.pola.rs/py-polars/html/reference/series/api/polars.Series.to_frame.html#polars.Series.to_frame) which makes the conversion explicit and allows for naming the series. I'm not sure an implicit conversion is good for the API.

**Comment by mcrumiller** [comment]
@mickvangelderen I'm not sure I follow; Series in principle should be interchangeable with eager DataFrame columns, and `hstacking` a (named) Series onto an existing df makes perfect logical sense:

```python
df = pl.DataFrame({
    "a": [1, 2, 3],
    "b": [4, 5, 6],
})
s = pl.Series("c", [7, 8, 9])

# current, we must df.hstack(s.to_frame())
print(df.hstack(s))
```
shape: (3, 3)
┌─────┬─────┬─────┐
│ a   ┆ b   ┆ c   │
│ --- ┆ --- ┆ --- │
│ i64 ┆ i64 ┆ i64 │
╞═════╪═════╪═════╡
│ 1   ┆ 4   ┆ 7   │
│ 2   ┆ 5   ┆ 8   │
│ 3   ┆ 6   ┆ 9   │
└─────┴─────┴─────┘
```

**Comment by mickvangelderen** [comment]
It is not immediately clear to me what "Combine multiple DataFrames, LazyFrames, or Series into a single object." means exactly in the [`concat` docs](https://docs.pola.rs/py-polars/html/reference/api/polars.concat.html). The type of items is `items: Iterable[PolarsType],` where `PolarsType = TypeVar("PolarsType", "DataFrame", "LazyFrame", "Series", "Expr")`. Does that mean that each item in the iterable has to be of the same concrete type? That would mean that you can concat a `DataFrame` with a `DataFrame`, and a `Series` with a `Series`, but not necessarily a `DataFrame` with a `Series`.

**Comment by stinodego** [comment]
> I'm marking this as an enhancement rather than as a bug since the docs clearly state "Series only support the vertical strategy." for `pl.concat`.
> 
> That said, I personally would be okay with allowing _named_ series on `pl.concat` for horizontal.
> 
> @stinodego What do you think?

I don't see why we couldn't support horizontal concatenation of Series. The user must make sure the Series names are unique, otherwise we raise an error.

**Comment by mcrumiller** [comment]
@mickvangelderen I agree that it's ambiguous, we should rework the language on that, although it will depend on the decision made in this issue.

I am not sure about mixing eager and lazy frames. We do allow for mixing lazy frames with series, as the Series is simply considered as input into the lazy query plan:

```python
>>> pl.LazyFrame().with_columns(
    pl.Series("a", [1, 2, 3])
).collect()
shape: (3, 1)
┌─────┐
│ a   │
│ --- │
│ i64 │
╞═════╡
│ 1   │
│ 2   │
│ 3   │
└─────┘
```
However, `join`, `concat`, `hstack`, etc. do not work with eager and lazy frame combinations. I feel that allowing joint operations on lazy and eager dataframes opens a Pandora's box that we should leave closed.
